File: BillCef/BillCefWeb/WebMain/DataHandler/MilitaryInvestigation.js
/// <reference path="../docs/carpa-vsdoc.js" />

Type.registerNamespace('WebMain.DataHandler');

WebMain.DataHandler.MilitaryInvestigationAction = function () {
    WebMain.DataHandler.MilitaryInvestigationAction.initializeBase(this);
    this.detailArray = {};
    this.socket = null;
}

WebMain.DataHandler.MilitaryInvestigationAction.prototype = {
    initialize: function () {
        WebMain.DataHandler.MilitaryInvestigationAction.callBaseMethod(this, 'initialize');
        this.doDataFormLoaded(this.get_form());
    },

    dispose: function () {
        WebMain.DataHandler.MilitaryInvestigationAction.callBaseMethod(this, 'dispose');
    },

    doSearch: function (sender) {
        var form = sender.get_form();

        var condition = form.condition.get_value();
        if (!condition || condition.length == 0) {
            alert('请录入侦查条件');
            return;
        }
        var data = JsObj.getMilitary(condition);

        var tempData = JSON.parse(data);
        for (var i in tempData) {
            var item = tempData[i];
            item.pic_url = "https:" + tempData[i].pic_url;
            item.buttons = "<font color='#999999' style='cursor:pointer;' title='查看详情'>查看详情</font>";
        }
        form.grid.dataBind(tempData);
    },

    doDataFormLoaded: function (sender) {
        var form = sender;
        var socket = this._connectWebsocket(sender);
        var t = new Date().getTime();
        this.uid = form.user.get_value() + "_" + t;
        socket.emit("login", { uid: this.uid })
    },

    _connectWebsocket: function (form) {
        var url = form.socketurl.get_value();
        if (url.length == 0) {
            alert('未设置socket对应的地址');
            return null;
        }
        if (this.socket) {
            return this.socket;
        }
        var _this = this;
        this.socket = io.connect(url);
        this.socket.on("goodMsg", function (data) {
            var items = {};
            try {
                items = JSON.parse(data.msg);
            } catch (e) {
                alert(data.msg);
                return;
            }

            if (items.gid) {//单商品
                var id = items.gid.toString();
                var data = form.grid.get_dataSource();
                var index = form.grid.get_selectedRowIndex();
                for (var i in data) {
                    var row = data[i];
                    if (!row.id) {
                        row.id = _this.getGoodId(row.detail_url);
                    }
                    if (row.id.toString() == id) {
                        row.start = items.startDate;
                        row.end = items.endDate;
                        var cache = _this.detailArray[row.detail_url];
                        if (cache) {
                            cache.Start = row.start;
                            cache.End = row.end;
                        }
                    }
                    if (i == index) {
                        _this.doGridChange(_this);
                    }
                }
                return;
            }
            for (var i in items) {
                var item = items[i];
                item.raw_title = item.title;
                item.start = item.startDate;
                item.end = item.endDate;
                item.create = item.create;
                item.detail_url = item.pic_url = "https:" + item.url;
                item.buttons = "<font color='#999999' style='cursor:pointer;' title='查看详情'>查看详情</font>";
            }
            form.grid.dataBind(items);
        });

        return this.socket;
    },

    doRefresh: function (sender) {
        var form = sender.get_form();
        form.refresh();
    },

    doGridChange: function (sender) {
        var form = sender.get_form();
        var data = form.grid.get_selectedRowData();
        if (!data || !data.detail_url) {
            return;
        }
        var detail_url = data.detail_url;
        var detail = this.detailArray[detail_url];
        if (!detail) {
            detail = JsObj.getDetail(data.detail_url);
            this.getGoodMsg(form, detail_url);
            detail = JSON.parse(detail);
            this.detailArray[detail_url] = detail;
        }

        if (detail.error) {
            delete this.detailArray[detail_url];
            console.log(detail.error);
            return;
        }
        form.gridSku.dataBind(detail.List);
        form.fhd.set_text(detail.SendCity);
        form.qrl.set_text(detail.ConfirmGoodsCount);
        form.xsl.set_text(detail.SoldTotalCount);
        form.zffs.set_text(detail.Pays);
        form.fw.set_text(detail.Service);
        form.yhj.set_text(detail.Coupon);
        form.start.set_text(detail.Start);
        form.end.set_text(detail.End);
    },

    doButtonClick: function (sender) {
        var form = sender.get_form();
        var data = form.grid.get_selectedRowData();
        if (!data || !data.detail_url) {
            return;
        }
        var detail_url = data.detail_url;
        window.open(detail_url, "_blank")
    },

    doGet: function (sender) {
        var form = sender.get_form();
        var value = form.saveData();
        this.getGoodMsg(form, value.condition);
    },

    getGoodMsg: function (form, condition) {
        var id = this.getGoodId(condition);
        if (id.length > 0) {
            form.get_service().GetSingleGoodMsg(id, this.uid, function () {

            });
        } else {
            form.get_service().GetGoodMsg(condition, this.uid, function () {
            });
        }
    },

    getGoodId: function (condition) {
        var items = /id=[0-9]+/.exec(condition);
        return items && items.length == 1 ? items[0].substr(3) : "";
    }
}
WebMain.DataHandler.MilitaryInvestigationAction.registerClass('WebMain.DataHandler.MilitaryInvestigationAction', Sys.UI.PageAction);

File: BillCef/BillCefWeb/WebMain/DataHandler/MilitaryInvestigation.cs
using System;
using Carpa.Web.Script;
using Carpa.Web.Ajax;
using Common;
using System.Web;
using System.Collections;
using WebMain.Common;

namespace WebMain.DataHandler
{
    [NeedLogin]
    public class MilitaryInvestigation : Page
    {
        string taodaxingUrl = "https://www.taodaxiang.com/shelf/index/get";
        string contentType = "application/x-www-form-urlencoded; charset=UTF-8";
        public override void Initialize()
        {
            base.Initialize();
            UserInfo info = (UserInfo)Session["user"];
            Context["user"] = info.User;
            Context["socketurl"] = IOUtils.SocketIOUrl;
        }

        [WebMethod]
        public void GetGoodMsg(string name, string uid)
        {
            JavaScriptSerializer serializer = JavaScriptSerializer.CreateInstance();

            HashObject data = new HashObject();
            data["fid"] = uid;
            data["url"] = string.Format("wwid={0}", HttpUtility.UrlEncode(name));
            SendMsg msg = new SendMsg("server");
            msg.comefrom = "net";
            msg.msg = serializer.Serialize(data);
            IOUtils.Emit("getGoodMsg", serializer.Serialize(msg));


            //JavaScriptSerializer serializer = JavaScriptSerializer.CreateInstance();

            //IHashObjectList list = new HashObjectList();
            //string value = Net.GetNetData(taodaxingUrl, string.Format("pattern=0&wwid={0}&goodid=&page=1", HttpUtility.UrlEncode(name)), contentType);
            //HashObject hash = serializer.Deserialize<HashObject>(value);
            //int code = hash.GetValue<int>("code");
            //if (code == 1)
            //{
            //    throw new Exception(string.Format("没有查询到【{0}】的网店信息", name));
            //}
            //CopyDataToList(hash, list);
            //while (code == 0) {
            //    var post = hash.GetValue<HashObject>("post");
            //    value = Net.GetNetData(taodaxingUrl, string.Format("pattern={0}&id={1}&page={2}&nt={3}", post["pattern"], post["id"], post["page"], post["nt"]), contentType);
            //    hash = serializer.Deserialize<HashObject>(value);
            //    code = hash.GetValue<int>("code");
            //    if (code != 0)
            //    {
            //        break;
            //    }
            //    CopyDataToList(hash, list);
            //}
            //return list;
        }

        [WebMethod]
        public void GetSingleGoodMsg(string id, string uid)
        {
            JavaScriptSerializer serializer = JavaScriptSerializer.CreateInstance();

            HashObject data = new HashObject();
            data["fid"] = uid;
            data["url"] = string.Format("goodid={0}", id);
            SendMsg msg = new SendMsg("server");
            msg.comefrom = "net";
            msg.msg = serializer.Serialize(data);
            IOUtils.Emit("getGoodMsg", serializer.Serialize(msg));

            //IHashObjectList list = new HashObjectList();
            //string value = Net.GetNetData(taodaxingUrl, string.Format("pattern=1&wwid=2&goodid={0}&page=1", id), contentType);
            //HashObject hash = serializer.Deserialize<HashObject>(value);
            //if (hash.GetValue<int>("code") == 0)
            //{
            //    CopyDataToList(hash, list);
            //    return list[0];
            //}
            //return value;
        }

        private void CopyDataToList(HashObject data, IHashObjectList list)
        {
            foreach (HashObject item in data.GetValue<ArrayList>("data"))
            {
                HashObject copy = new HashObject();
                copy.Add("gid", item["goodid"]);
                copy.Add("title", item["title"]);
                copy.Add("url", item["url"]);
                copy.Add("startDate", string.Format("{0}000", item["start"]));
                copy.Add("endDate", string.Format("{0}000", item["end"]));
                copy.Add("create", string.Format("{0}000", item["create"]));
                list.Add(copy);
            }
        }
    }
}
